chore: migrate swaps to BIP-44

bfullam opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**
Integrating the new BIP-44 account selectors into the Swaps/Bridge experience.

### Added

- **Source account selector**:  a new `selectSourceWalletAddress` selector has been added to the bridge store, which wraps `selectSelectedInternalAccountByScope`.
- **Destination account selector component**: filter by scope instead of explicit Solana checks.
- **Display the AccountGroup's name instead of the individual account's name**: check new selector `selectAccountGroups` for group name.

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry:

## **Related issues**

Fixes:

## **Manual testing steps**

```gherkin
Feature: BIP-44 selectors

  Scenario: user completes EVM <> NON-EVM swap
    Given user has an asset to swap

    When user inputs swap parameters
    Then proper destination account options are available and swap completes as expected
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

bfullam committed:
chore: add account group naming to useAccounts

bfullam committed:
chore: filter accounts by scope

bfullam committed:
chore: migrate from selectSelectedInternalAccountAddress to selectSourceWalletAddress

bfullam committed:
chore: remove explicit solana checks

bfullam committed:
test: add selectSourceWalletAddress mocks to tests

bfullam committed:
test: fix destinationAccountSelector tests

bfullam committed:
test: update bridgeView tests

bfullam committed:
Merge branch 'main' into swaps-2788-update-selectors-to-BIP44

bfullam committed:
Merge branch 'main' into swaps-2788-update-selectors-to-BIP44

bfullam committed:
chore: gate account names behind feature flag

owencraston commented on app/components/hooks/useAccounts/useAccounts.ts:
```diff
@@ -166,10 +172,18 @@ const useAccounts = ({
+        let accountName = internalAccount.metadata.name;
+        if (isMultichainAccountsState2Enabled) {
+          accountName =
+            accountGroups.find(({ accounts: groupAccounts }) =>
```
I really want to avoid modifying the useAccounts hook since its already so dam slow and we have had countless performance issues with it. This hook is used a lot too so the changes here have wide spread effects. Let me see if I can find a better way.

ccharly replied:
Also, why would we need to rely on `useAccounts` anymore? 🤔

We're shifting the entire account-model from singular blockchain accounts to "account groups", so I'd say the components that are being migrated to "multichain accounts" have to use new selectors and not rely on the existing selectors/hooks for the old account model?

I do miss some context here maybe, but we should not "fix"/"alter" the old behaviors, we need to re-write/migrate the components that have to be compatible with "multichain accounts"/BIP-44.

owencraston committed:
try to improve solution

owencraston commented:
I've made some very crude modifications on this [draft PR](https://github.com/MetaMask/metamask-mobile/pull/18909) take a look and let me know what you think. My main goal was to remove the added logic from useAccounts.

ccharly commented on app/components/UI/Bridge/components/DestinationAccountSelector.tsx/index.tsx:
```diff
@@ -57,6 +61,21 @@ const DestinationAccountSelector = () => {
+    const destCaipChainId = formatChainIdToCaip(destToken.chainId);
+    const isEvmChain = !isNonEvmChainId(destCaipChainId);
+
+    return accounts.filter(
+      (account) =>
+        account.scopes?.includes(destCaipChainId) ||
+        (isEvmChain && account.scopes?.includes(EthScope.Eoa)),
+    );
```
We should use `isScopeEqual` and `isScopeEqualToAny` utils for those (if we decide to keep the filtering here of course):

Reference is here:
- https://github.com/MetaMask/accounts/blob/main/packages/keyring-utils/src/scopes.ts

bfullam committed:
chore: use account group naming when available

bfullam committed:
feat: try to improve solution (#18909)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the
CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing
description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and
accurately)
-->

CHANGELOG entry:

## **Related issues**

Fixes:

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

bfullam committed:
chore: remove useAccounts changes

bfullam committed:
Merge branch 'main' into swaps-2788-update-selectors-to-BIP44

owencraston commented on app/selectors/bridge.ts:
```diff
@@ -0,0 +1,51 @@
+
+    const byDestScope = selectInternalAccountsByScope(state, destScope);
+    const evmWildcard = isEvm
+      ? selectInternalAccountsByScope(state, 'eip155:0')
```
nit: we should use EthScope.EOA here.

owencraston reviewed (commented):
With BIP44 all account should support solana so when I perform a swap while my current selected account is account three, I would expect the default swap/bridge destination account to be my current selected account. Here you can see in my screenshots, the default destination is not my current selected account. Another weird thing is that there are only 3 valid destination addresses. I think all account should support solana but I think these are missing due to my team not finishing the account reconciliation yet. 

<img width="250" height="600" alt="Simulator Screenshot - iPhone 16 Plus - 2025-08-29 at 13 49 39" src="https://github.com/user-attachments/assets/d684c640-df1c-49fc-b865-3baf14f024bc" />

<img width="250" height="600" alt="Simulator Screenshot - iPhone 16 Plus - 2025-08-29 at 13 49 32" src="https://github.com/user-attachments/assets/85d576d1-4023-4aff-b577-12cd33e75512" />

<img width="250" height="600" alt="Simulator Screenshot - iPhone 16 Plus - 2025-08-29 at 13 50 01" src="https://github.com/user-attachments/assets/678f0354-dd72-4d1d-9d53-1b57b8ad6d70" />

<img width="250" height="600" alt="Simulator Screenshot - iPhone 16 Plus - 2025-08-29 at 13 50 14" src="https://github.com/user-attachments/assets/5fe78999-ebfd-4228-8c76-16d4edfda78f" />

bfullam committed:
chore: use EthScope.Eoa instead of string literal

bfullam committed:
feat: use currently selected account group as default destination account if possible

bfullam committed:
test: add account group tests

bfullam committed:
test: add mock state

bfullam committed:
Update destination account picker icon

bfullam committed:
test: update test

bfullam committed:
Merge branch 'main' into swaps-2788-update-selectors-to-BIP44

bfullam committed:
Merge branch 'main' into swaps-2788-update-selectors-to-BIP44

bfullam committed:
Merge branch 'main' into swaps-2788-update-selectors-to-BIP44
